File: analysis/parse_args.py
# This script sets up the configuration for the Python analysis pipeline.
# It reads in user command-line arguments and then extracts the appropriate
# configuration settings from config.json based on those arguments.

import argparse
import json

# Load default config from JSON
with open("analysis/config.json", "r") as f:
    config = json.load(f)

parser = argparse.ArgumentParser()  # Instantiate parser

# ----------------- Parse user arguments -------------------------------

# Configuration for add measures
parser.add_argument(
    "--add_indicat_prescript",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Adds indicat/prescript if flag is added to action.",
)
parser.add_argument(
    "--add_prescriptions",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Adds prescriptions if flag is added to action.",
)
parser.add_argument(
    "--add_reason",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Adds reason if flag is added to action.",
)
parser.add_argument(
    "--demograph_measures",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Sets measures defaults to demographic-level subgroups",
)
parser.add_argument(
    "--practice_measures",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Sets measures defaults to practice-level subgroups",
)
parser.add_argument(
    "--comorbid_measures",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Sets measures defaults to comorbidity-level subgroups",
)
parser.add_argument(
    "--practice_subgroup_measures",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Sets measures defaults to practice subgroup-level subgroups",
)
parser.add_argument(
    "--use_csv",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Uses csv instead of arrow for ease of file inspection",
)
parser.add_argument(
    "--set",
    default=argparse.SUPPRESS,
    help="Choose which set of measures to extract: appts_table, sro, resp",
)

# Configuration for interval date input
parser.add_argument(
    "--start_intv", default=argparse.SUPPRESS, help="Interval start date"
)
parser.add_argument(
    "--yearly",
    action = "store_true",
    default=argparse.SUPPRESS,
    help="Set intervals to yearly instead of weekly",
)
parser.add_argument(
    "--weekly_agg",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Set yearly intervals to be aggregated from weekly measures",
)
parser.add_argument(
    "--test",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Sets lowest time intervals for lightweight testing",
)

# Restrict measure to those with an appt
parser.add_argument(
    "--appt",
    action="store_true",
    default=argparse.SUPPRESS,
    help="Restrict measures to those with an appointment in interval",
)

args = parser.parse_args()  # Stores arguments in 'args'

# Override config with provided args
for key, value in vars(args).items():
    config[key] = value

# ----------------- Apply conditional logic to config -------------------

# Initialize dtype_dict with base
config["dtype_dict"] = config["base_dtype_dict"].copy()

# Apply group-specific configuration
for group in ["demograph", "practice", "comorbid", "practice_subgroup"]:
    if config.get(f"{group}_measures", False):
        # Set group in config and update dtype_dict
        config["group"] = group
        config["dtype_dict"].update(config["groups"][group]["dtype_dict"])
        break  # Only one group can be selected

if config.get("appt", False):
    config["appt_suffix"] = "_appt"

if config.get("weekly_agg", False):
    config["agg_suffix"] = "_weeklyagg"

if config.get("test", False):
    config["test_suffix"] = "_test"

config["deprioritized"] = set(config["sro_dict"].keys()) - set(config["prioritized"])

if config.get("use_csv", False):
    config["file_type"] = "csv"

if config.get("set") == "sro":
    config["pipeline_measures"] = config["measures_list"]["sro"]
elif config.get("set") == "resp":
    config["pipeline_measures"] = config["measures_list"]["resp"]
elif config.get("set") == "appts_table":
    config["pipeline_measures"] = config["measures_list"]["appts_table"]

File: analysis/parse_args.r
# This script sets up the configuration for the R analysis pipeline.
# It reads in user command-line arguments and then extracts the appropriate
# configuration settings from config.json based on those arguments.

library(optparse)
library(styler)
library(jsonlite)

# Run >R config.r to lint repo
#styler::style_dir("analysis")

# Load config
config <- fromJSON(file.path(dirname(sys.frame(1)$ofile), "config.json"))

# ----------------- Parse user arguments -------------------------------

# Define list of options
option_list <- list(
  make_option("--add_indicat_prescript",
    action = "store_true",
    default = config$add_indicat_prescript, help = "Adds indicat/prescript if flag is added to action."
  ),
  make_option("--add_prescriptions",
    action = "store_true",
    default = config$add_prescriptions, help = "Adds prescriptions if flag is added to action."
  ),
  make_option("--add_reason",
    action = "store_true",
    default = config$add_reason, help = "Adds reason if flag is added to action."
  ),
  make_option("--demograph_measures",
    action = "store_true",
    default = config$demograph_measures, help = "Sets measures defaults to demographic-level subgroups."
  ),
  make_option("--practice_measures",
    action = "store_true",
    default = config$practice_measures, help = "Sets measures defaults to practice-level subgroups."
  ),
  make_option("--comorbid_measures",
    action = "store_true",
    default = config$comorbid_measures, help = "Sets measures defaults to comorbidity-level subgroups."
  ),
  make_option("--practice_subgroup_measures",
    action = "store_true",
    default = config$practice_subgroup_measures, help = "Sets measures defaults to practice subgroup-level subgroups."
  ),
  make_option("--start_intv",
    type = "character", default = config$start_intv,
    help = "Interval start date."
  ),
  make_option("--test",
    action = "store_true",
    default = config$test, help = "Sets lowest time intervals for lightweight testing."
  ),
  make_option("--use_csv",
    action = "store_true", default = config$use_csv,
    help = "Use CSV files instead of Arrow for reading/writing data."
  ),
  make_option("--set",
    type = "character",
    default = config$set, help = "Choose set of measures between 1) all 2) sro 3) resp."
  ),
  make_option("--appt",
    action = "store_true",
    default = config$appt, help = "Restrict measures to those with an appointment in interval"
  ),
  make_option("--yearly",
    action = "store_true",
    default = config$yearly, help = "Use yearly measures data"
  ),
  make_option("--released", 
    action = "store_true", 
    default = FALSE, help = "Use already released data"
  ),
  make_option("--weekly_agg", 
    action = "store_true", 
    default = config$weekly_agg, help = "Generate weekly aggregates"
  )
)

# Parse options
parser <- OptionParser(option_list = option_list)
args <- parse_args(parser)

# Override config with provided args
args_list <- as.list(args)
for (key in names(args_list)) {
  config[[key]] <- args_list[[key]]
}

# ----------------- Apply conditional logic to config -------------------

# Configure dates (not needed in R)

# Initialize dtype_dict with base (not used in R, but for consistency)
config$dtype_dict <- config$base_dtype_dict

# Apply group-specific configuration
group_selected <- FALSE
for (group in c("demograph", "practice", "comorbid", "practice_subgroup")) {
  group_flag <- paste0(group, "_measures")
  if (config[[group_flag]]) {
    config$group <- group
    config$dtype_dict <- c(config$dtype_dict, config$groups[[group]]$dtype_dict)
    group_selected <- TRUE
    break  # Only one group can be selected
  }
}

if (config$appt) {
  config$appt_suffix <- "_appt"
} 

if (config$yearly) {
  if (config$weekly_agg) {
    config$agg_suffix <- "_weeklyagg"
  } else {
    config$agg_suffix <- "_yearly"
  }
}

if (config$test) {
  config$test_suffix <- "_test"
}

config$deprioritized <- setdiff(names(config$sro_dict), config$prioritized)

if (config$use_csv) {
  config$file_type <- "csv"
}

if (!is.null(config$set)) {
  if (config$set == "sro") {
    config$pipeline_measures <- config$measures_list$sro
  } else if (config$set == "resp") {
    config$pipeline_measures <- config$measures_list$resp
  } else if (config$set == "appts_table") {
    config$pipeline_measures <- config$measures_list$appts_table
  }
}
